pennylane/labs/intermediate_reps/rowcol.py

```diff
@@ -0,0 +1,214 @@
+    """In-place apply update corresponding to a CNOT on wires ``(control, target)``
+    to parity matrix P and list of CNOT gates ``cnots``."""
+    P[target] += P[control]
+    cnots.append((control, target))
+
+
+F_2 = galois.GF(2)
+
+
+def _get_S(P: np.ndarray, idx: int, node_set: Iterable[int], mode: str):
+    # Find S (S') either by simply extracting a column or by solving a linear system for the row
+    if mode == "column":
+        b = P[:, idx]
+    else:
+        P = F_2(P)
+        e_i = F_2.Zeros(len(P))
+        e_i[idx] = 1
+        b = np.linalg.solve(P.T, e_i)  # This solve step is over F_2!
+    S = set(np.where(b)[0])
+    # Add the node ``idx`` itself
+    S.add(idx)
+    # Manually remove nodes from S that are no longer part of ``connectivity``.
+    S = S.intersection(node_set)
+    return S
+
+
+def _eliminate(P: np.ndarray, connectivity: nx.Graph, idx: int, mode: str, verbose: bool):
+    """Eliminate the column or row with index ``idx`` of the parity matrix P,
+    respecting the connectivity constraints given by ``connectivity``.
+
+    Args:
+        P (np.ndarray): Parity matrix
+        connectivity (nx.Graph): Connectivity graph
+        idx (int): Column or row index to eliminate
+        mode (str): Whether to eliminate the column (``column``) or row (``row``) of ``P``.
+        verbose (bool): Whether to print elimination results
+
+    Returns:
+        tuple[np.ndarray, list[tuple[int]]]: Updated parity matrix and list of CNOTs that
+        accomplish the update, in terms of ``(control, target)`` qubit pairs.
+    """
+
+    # i.1.1/i.2.1 Construct S (mode="column") or S' (mode="row"), respectively.
+    S = _get_S(P, idx, connectivity.nodes(), mode)
+    if len(S) == 1:
+        # idx is in S for sure, so this column/row already has the right content. No need to update
+        if verbose:
+            print(f"CNOTs from {mode} elimination ({idx}): {[]}")
+        return P, []
+
+    cnots = []
+    # i.1.1/i.2.1 Find Steiner tree within S (S').
+    T = steiner_tree(connectivity, list(S))
+
+    # Need post-order nodes in any case
+    visit_nodes = postorder_traverse(T, source=idx)
+
+    if mode == "column":
+        # For column mode use post-order and parities constraint in first pass (i.1.2)...
+        _ = [
+            _update(P, cnots, child, parent)
+            for child, parent in visit_nodes
+            if P[child, idx] == 1 and P[parent, idx] == 0
+        ]
+        # ... and no constraints for second pass (i.1.3)
+        _ = [_update(P, cnots, parent, child) for child, parent in visit_nodes]
+
+    else:
+        # For row mode use pre-order and constraints from S' in first pass (i.2.2)...
+        previsit_nodes = preorder_traverse(T, source=idx)
+        _ = [_update(P, cnots, child, parent) for child, parent in previsit_nodes if child not in S]
+        # ... and no constraints for second pass (i.2.3)
+        _ = [_update(P, cnots, child, parent) for child, parent in visit_nodes]
```

Comment:
as mentioned above, a bit sketchy with all the inplace and global context operations. if you can think of a way to make it more proper without destroying the neatness of this it'd be nice to have

Reply:
Yeah, so I pushed a new version.
I think I personally like the previous one better, but I understand the dislike of inplace mutation,as it is a bit old school and less readable/maintainable/stable.